app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+require 'rqrcode'
+
+module Users
+  class NbpWalletController < ApplicationController
+    skip_after_action :verify_authorized
+
+    def connect
+      if Enmeshed::Connector.pending_relationship_for_nbp_uid(@provider_uid).present?
+        redirect_to nbp_wallet_finalize_users_path and return
+      end
+
+      @template = Enmeshed::Connector.create_relationship_template(@provider_uid)
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data Enmeshed::RelationshipTemplate.new(params[:truncated_reference]).qr_code, type: 'image/png'
+    end
+
+    def relationship_status
+      if Enmeshed::Connector.pending_relationship_for_nbp_uid(@provider_uid).present?
+        render json: {status: :ready}
+      else
+        render json: {status: :waiting}
+      end
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      @relationship = Enmeshed::Connector.pending_relationship_for_nbp_uid(@provider_uid)
+      abort_and_refresh and return if @relationship.blank?
+
+      accept_and_create_user
+    rescue Enmeshed::ConnectorError, Faraday::Error => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    def cancel
+      session.delete(:saml_uid)
+      session.delete(:omniauth_provider)
```

Comment:
This one lacks the removal of `saml_session_index`. However, this method might become obsolete when switching to the existing logout workflow.